Source/JavaScriptCore/b3/air/AirOptimizePairedLoadStore.cpp

```diff
@@ -0,0 +1,311 @@
+        return Width64;
+    default:
+        RELEASE_ASSERT_NOT_REACHED();
+        return Width8;
+    }
+}
+
+static bool tryStorePair(Code& code, BasicBlock* block, unsigned current, Inst& inst)
+{
+    Width instWidth = accessWidth(inst.kind.opcode);
+    int64_t instOffset = static_cast<int64_t>(inst.args[1].offset());
+    unsigned limit = std::min(current, AirOptimizePairedLoadStoreInternal::scanInstructions);
+    RegisterSet clobbered;
+    for (unsigned count = 1; count <= limit; ++count) {
+        unsigned index = current - count;
+        Inst& target = block->at(index);
+
+        auto logFound = [&](const Inst& newInst) {
+            if (AirOptimizePairedLoadStoreInternal::verbose) {
+                dataLogLnIf(AirOptimizePairedLoadStoreInternal::verbose, "FOUND  ", inst, " ", target, " => ", newInst);
+                for (unsigned i = 1; i < count; ++i)
+                    dataLogLnIf(AirOptimizePairedLoadStoreInternal::verbose, "    ", block->at(current - i));
+            }
+        };
+
+        auto logFailed = [&]() {
+            if (AirOptimizePairedLoadStoreInternal::verbose) {
+                dataLogLnIf(AirOptimizePairedLoadStoreInternal::verbose, "FAILED ", inst, " ", target);
+                for (unsigned i = 1; i < count; ++i)
+                    dataLogLnIf(AirOptimizePairedLoadStoreInternal::verbose, "    ", block->at(current - i));
+            }
+        };
+
+        // If the instruction has some special effect (including Patchpoint), we give up since we cannot model the effect of this.
+        if (target.hasNonArgEffects()) {
+            logFailed();
+            return false;
+        }
+
+        // If some instructions between the current and target clobbers registers used for current / target,
+        // then we cannot merge them since the current instruction's registers are changed.
+        //
+        //     stur %x1, [%fp]
+        //     movz %x2, #0
+        //     stur %x2, [%fp, #8]
+        //
+        // Then we cannot make it to
+        //
+        //     stp %x1, %x2, [%fp]
+        //     movz %x2, #0
+        Inst::forEachDefWithExtraClobberedRegs<Tmp>(
+            &block->at(index), &block->at(index + 1),
+            [&] (const Tmp& arg, Arg::Role, Bank, Width, PreservedWidth) {
+                clobbered.add(arg.reg(), IgnoreVectors);
+            });
+        if (clobbered.contains(inst.args[1].base().reg(), IgnoreVectors) || (inst.args[0].isTmp() && clobbered.contains(inst.args[0].reg(), IgnoreVectors))) {
+            logFailed();
+            return false;
+        }
+
+        {
+            // If some instructions between the current and target have memory-load and memory-store,
+            // then we cannot merge them since reordering can change the results.
+            // But this is really pessimistic: if the base is the same to the current instruction, and if the offset
+            // is different from the current instruction, it is OK actually.
```

Comment:
This comment seems outdated? Also, it seems like it should be ok to reorder 3 loads to make a loadPair. Maybe we should comment on that.

Reply:
This part is only for StorePair (this function only do store merging), so I think this is correct.

Reply:
Ah right, good point.